File: _modules/ssha.py
#!/usr/bin/env python

# http://www.openldap.org/faq/data/cache/347.html

import os
import hashlib
from base64 import encodestring as encode
from base64 import decodestring as decode

def b64e(string):
    return encode(string)[:-1]

def makeSecret(password,salt=None):
    if salt is None:
        salt = os.urandom(10)
    h = hashlib.sha1(password)
    h.update(salt)
    return "{SSHA}" + encode(h.digest() + salt)[:-1]

def checkPassword(challenge_password, password):
    challenge_bytes = decode(challenge_password[6:])
    digest = challenge_bytes[:20]
    salt = challenge_bytes[20:]
    hr = hashlib.sha1(password)
    hr.update(salt)
    return digest == hr.digest()

File: openldap/files/ssha.rb
#!/usr/bin/env ruby

# http://www.openldap.org/faq/data/cache/347.html

# Ruby script to generate SSHA (Good for LDAP)
begin
  require 'sha1'
  require 'base64'
  require 'io/console'
rescue LoadError
end

if STDIN.respond_to?(:noecho)
  def get_password(prompt="Password: ")
    print prompt
    STDIN.noecho(&:gets).chomp
  end
else
  def get_password(prompt="Password: ")
    `read -s -p "#{prompt}" password; echo $password`.chomp
  end
end

password1 = get_password("Enter your password here   : ")
puts ''
password2 = get_password("Re-enter your password here: ")
puts ''

require 'securerandom'
salt = SecureRandom.hex

if password1 != password2
  puts "Passwords don't match!"
else
  hash = "{SSHA}"+Base64.encode64(Digest::SHA1.digest(password2+salt)+salt).chomp!
  puts hash+"\n"
end
